compiler/src/iree/compiler/GlobalOptimization/DemoteContractionInputsToBF16.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+namespace {
+
+template <typename OpType>
+void rewriteOpWithNewInputs(PatternRewriter &rewriter, linalg::LinalgOp op,
+                            ValueRange newInputs) {
+  auto namedOp = cast<OpType>(op);
+  rewriter.replaceOpWithNewOp<OpType>(op, newInputs, op.getDpsInits(),
+                                      linalg::getPrunedAttributeList(namedOp));
+}
+
+// For narrowable inputs, selects
+struct DemoteContractionInputsToBF16Pattern
+    : public OpInterfaceRewritePattern<linalg::ContractionOpInterface> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+  LogicalResult matchAndRewrite(linalg::ContractionOpInterface op,
+                                PatternRewriter &rewriter) const override {
+    linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(op.getOperation());
+    for (auto operand : linalgOp->getOperands()) {
+      auto operandType = dyn_cast<RankedTensorType>(operand.getType());
+      if (!operandType ||
+          operandType.getElementType() != rewriter.getF32Type()) {
+        return failure();
+      }
+    }
+    Location loc = linalgOp.getLoc();
+    SmallVector<Value> demotedInputs;
+    for (auto inputOperand : linalgOp.getDpsInputOperands()) {
+      auto input = inputOperand->get();
+      auto inputType = cast<RankedTensorType>(input.getType());
+      auto demotedInputType =
+          RankedTensorType::get(inputType.getShape(), rewriter.getBF16Type(),
+                                inputType.getEncoding());
+      SmallVector<AffineMap> maps(
+          2, AffineMap::getMultiDimIdentityMap(inputType.getRank(),
+                                               rewriter.getContext()));
```

Comment:
nit: we can use `rewriter.getMultiDimIdentityMap(inputType.getRank())`, then we don't need the MLIRcontext.